### ISSUE TITLE
Parallelisation with python2 kernel while running Jupyterhub with python3

### ISSUE THREAD

**Issue by precicely** [ipython/ipyparallel#131]
Good day,

I was hoping someone could help me diagnose the following problem:

I'm running Jupyterhub from a python 3 Anaconda environment. However, I seem to have trouble running pyparallel code in a python 2 kernel within the aforementioned Jupyterhub (and hence python3) environment and I've come a cross what I believe to be a pickling issue but I may well be wrong...

When I run the following (with with a python3 kernel):

``` python
import ipdb
from ipyparallel import Client
c = Client()
print(c.ids)
print(c[:].apply_sync(lambda : "Hello, World"))
```

I get this:

```
[0, 1]
['Hello, World', 'Hello, World']
```

Which is what we expect.

However, when doing the same with a python 2 kernel:

``` python
import ipdb
from ipyparallel import Client
c = Client()
print c.ids
print c[:].apply_sync(lambda : "Hello, World")
```

I get the following:

```
[0, 1]
```

```
>[Engine Exception]
>--------------------------------------------------------------------------->TypeError                                 Traceback (most recent call last)/opt/anaconda3/envs/py34/lib/>python3.4/site-packages/ipyparallel/serialize/serialize.py in unpack_apply_message(bufs, g, copy)
>    163     assert len(bufs) >= 2, "not enough buffers!"
>    164     pf = buffer_to_bytes_py2(bufs.pop(0))
>--> 165     f = uncan(pickle.loads(pf), g)
>    166     pinfo = buffer_to_bytes_py2(bufs.pop(0))
>    167     info = pickle.loads(pinfo)
>/opt/anaconda3/envs/py34/lib/python3.4/site-packages/ipyparallel/serialize/codeutil.py in code_ctor(*args)
>     22 
>     23 def code_ctor(*args):
>---> 24     return types.CodeType(*args)
>     25 
>     26 def reduce_code(co):
>TypeError: an integer is required (got type str)

>[Engine Exception]
>--------------------------------------------------------------------------->TypeError                                 Traceback (most recent call last)/opt/anaconda3/envs/py34/lib/>python3.4/site-packages/ipyparallel/serialize/serialize.py in unpack_apply_message(bufs, g, copy)
>    163     assert len(bufs) >= 2, "not enough buffers!"
>    164     pf = buffer_to_bytes_py2(bufs.pop(0))
>--> 165     f = uncan(pickle.loads(pf), g)
>    166     pinfo = buffer_to_bytes_py2(bufs.pop(0))
>    167     info = pickle.loads(pinfo)
>/opt/anaconda3/envs/py34/lib/python3.4/site-packages/ipyparallel/serialize/codeutil.py in code_ctor(*args)
>     22 
>     23 def code_ctor(*args):
>---> 24     return types.CodeType(*args)
>     25 
>     26 def reduce_code(co):
>TypeError: an integer is required (got type str)
```

As I said above, I believe it has to do with pickling in the following:  ipyparallel/ipyparallel/serialize/codeutil.py lines 18-21. However, I'm no expert in these matters and would appreciate any help you could provide.

Cheers

**Comment by precicely** [comment]
Ok, so with some further digging I've discovered the following:

If I start the clusters manually in my **python 2** environment, like so:
>user@server:~$ source activate py27
>(py27)user@server:~$ ipcluster start --n=2

Then I run the  snippet below in Jupyter (remembering that its running a **python 2** kernel that's being hosted on Jupyterhub which in turn is running from a python 3 environment [apologies for the the Russian dolls of environments but bare with me!]):

`import ipdb`
`from ipyparallel import Client`
`c = Client()`
`print c.ids`
`print c[:].apply_sync(lambda : "Hello, World")`

I get the following:
>[0, 1]
>['Hello, World', 'Hello, World']

Which is what we expect!

So I believe the root issue lies with starting the kernels from the "IPython Clusters" tab in Jupyter. These kernels are python3 in nature and as such my prevously hypothesized issues with pickling. I checked this by doing the following (purposefully launching **python 3** clusters to see if it breaks):

>user@server:~$ source activate py34
>(py34)user@server:~$ ipcluster start --n=2

Then in Jupyter:

`import ipdb`
`from ipyparallel import Client`
`c = Client()`
`print c.ids`
`print c[:].apply_sync(lambda : "Hello, World")`

Which gives the same error as the one I reported in my previous post:

>[0, 1]
>
>[Engine Exception]
>--------------------------------------------------------------------------->TypeError Traceback (most recent call last)/opt/anaconda3/envs/py34/lib/>python3.4/site-packages/ipyparallel/serialize/serialize.py inunpack_apply_message(bufs, g, copy)
>163 assert len(bufs) >= 2, "not enough buffers!"
>164 pf = buffer_to_bytes_py2(bufs.pop(0))
>--> 165 f = uncan(pickle.loads(pf), g)
>166 pinfo = buffer_to_bytes_py2(bufs.pop(0))
>167 info = pickle.loads(pinfo)
>/opt/anaconda3/envs/py34/lib/python3.4/site-packages/ipyparallel/serialize/codeutil.py in code_ctor(args)
>22
>23 def code_ctor(args):
>---> 24 return types.CodeType(*args)
>25
>26 def reduce_code(co):
>TypeError: an integer is required (got type str)>
>[Engine Exception]
>--------------------------------------------------------------------------->TypeError Traceback (most recent call last)/opt/anaconda3/envs/py34/lib/>python3.4/site-packages/ipyparallel/serialize/serialize.py inunpack_apply_message(bufs, g, copy)
>163 assert len(bufs) >= 2, "not enough buffers!"
>164 pf = buffer_to_bytes_py2(bufs.pop(0))
>--> 165 f = uncan(pickle.loads(pf), g)
>166 pinfo = buffer_to_bytes_py2(bufs.pop(0))
>167 info = pickle.loads(pinfo)
>/opt/anaconda3/envs/py34/lib/python3.4/site-packages/ipyparallel/serialize/codeutil.py in code_ctor(args)
>22
>23 def code_ctor(args):
>---> 24 return types.CodeType(*args)
>25
>26 def reduce_code(co):
>TypeError: an integer is required (got type str)
I started the clusters manually in my **python 3** environment, like so:
>user@server:~$ source activate py34
>(py34)user@server:~$ ipcluster start --n=2

So it would appear that clusters launched from the Jupyter webpage inherit the environment of their host (hence they are not agnostic as I previously thought), however, it does mean that running parallel **python2** code in Jupyterhub is impossible as Jupyterhub requires python3.3 and above.

Does that make sense or am I way off the mark? If not, perhaps this post will help someone in the future...

TL;DR
if you want to run **python 2** code in Jupyterhub, launch your clusters from the command line in your python2 environment, like so:
>user@server:~$ source activate py27
>(py27)user@server:~$ ipcluster start --n=2

**Comment by minrk** [comment]
It looks like you have tracked it down: In general that the `ipcluster` command should be run with the same Python version as where you create the Client (your Kernel).

**Comment by precicely** [comment]
Should we raise this with the Jupyter folks? Perhaps where one specifies the "# of engines" in the "IPython Clusters" tab, one could also choose the type of clusters (i.e python version).

I imagine however, that its a corner-case too far...

**Comment by minrk** [comment]
No, it's appropriate to continue here. The clusters tab is in fact [provided by this repo](https://github.com/ipython/ipyparallel/tree/5.0.1/ipyparallel/nbextension) as a Jupyter extension. I think it's fine to add the Python version info to the cluster tab, given how it works today. I would like to change that in the future, but that would all be part of this repo.

**Comment by jonhilgart22** [comment]
I was wondering why my Python 2.x environment was not working with a Python 3.x cluster. Thansk for the help!

**Comment by rainwoodman** [comment]
Is this the proper interface to select a Jupyter kernel for the engine? It appears currently the cluster manager always launch with the default kernel? How do I switch the engines to use a different kernel?

**Comment by rainwoodman** [comment]
Looks like instead of using IPClusterStart directly, we can start a IPython kernel with the selected kernel add the IPClusterStart as a payload. If the kernel does not support python then this will die ungracefully.

**Comment by minrk** [comment]
To start an IPython engine, one launches `ipengine`, not a Jupyter kernel. The `python2` vs `python3` kernel distinction is a bit confusing because it suggests that ipyparallel uses Jupyter kernels, which isn't really true. IPython parallel engines extend the *IPython* kernel to produce an engine, but it is a distinct *Python-specific* object that is not compatible in general with how Jupyter kernels work.
